File: src/modules/notification.py

from .telegram import send as telegram_send
from .discord import send as discord_send

def send_notification (
	exit_code = None,
	log_file_name = None,
	repo_display_name = None,
	notifiers = None,
):
	is_ok = exit_code == 0

	if is_ok:
		text = '✅ Build passed'
	else:
		text = '❌ Build failed'

	text += f': *{ repo_display_name }*'
	text_logs = ''

	if not is_ok:
		with open(log_file_name, 'r') as file:
			last_lines_list = file.read().splitlines()[-10:]
			last_lines = '\n'.join(last_lines_list)
			text_logs = '\n\n*Exit code:* {}\n*Output log:*\n```\n{}```'.format(exit_code, last_lines)

	for notifier in notifiers:
		text_here = text
		log_file_name_here = None
		if not is_ok and 'logs' in notifier and notifier['logs'] == True:
			text_here += text_logs
			log_file_name_here = log_file_name

		if notifier['type'] == 'telegram':
			telegram_send(
				notifier = notifier,
				text = text_here,
				log_file_name = log_file_name_here,
			)
		elif notifier['type'] == 'discord':
			discord_send(
				notifier = notifier,
				text = text_here,
				log_file_name = log_file_name_here,
			)
		else:
			print('Unknown notifier type:', notifier['type'])

File: src/run/notification.js

import * as ARGS       from './args.js';
import * as loggerFile from './logger.file.js';
import sendDiscord	   from './notification/discord.js';
import sendTelegram    from './notification/telegram.js';

export default async function (exit_code) {
	const is_ok = exit_code === 0;

	let text = is_ok ? '✅ Build passed' : '❌ Build failed';
	text += `: *${ARGS.display_name}*`;

	let text_logs = '';
	if (!is_ok) {
		const log = await loggerFile.read();
		const log_last_lines = log.split('\n').slice(-10).join('\n');

		text_logs = '\n\n*Exit code:* ' + exit_code + '\n*Output log:*\n```\n' + log_last_lines + '```';
	}

	for (const notifier of ARGS.notifiers) {
		const add_logs = notifier.logs && !is_ok;
		const args = [
			notifier,
			text + (add_logs ? text_logs : ''),
			add_logs,
		];

		switch (notifier.type) {
			case 'telegram':
				await sendTelegram(...args); // eslint-disable-line no-await-in-loop
				break;
			case 'discord':
				await sendDiscord(...args); // eslint-disable-line no-await-in-loop
				break;
			// no default
		}
	}
}
